internal/agent/device/dependency/dependency.go

```diff
@@ -1,20 +1,406 @@
+			return fmt.Errorf("scheduling prefetch targets: %w", err)
+		}
+	}
+
+	return m.checkReady(ctx)
+}
+
+func (m *prefetchManager) checkReady(ctx context.Context) error {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	var pending []string
+	for image, task := range m.tasks {
+		if ctx.Err() != nil {
+			return ctx.Err()
+		}
+		if !task.done {
+			pending = append(pending, fmt.Sprintf("%s in progress", image))
+			continue
+		}
+		if task.err != nil {
+			if errors.IsRetryable(task.err) {
+				pending = append(pending, fmt.Sprintf("%s retrying: %v", image, task.err))
+			} else {
+				return task.err
+			}
+		}
+	}
+
+	if len(pending) > 0 {
+		return fmt.Errorf("%w: %v", errors.ErrPrefetchNotReady, pending)
+	}
+	return nil
+}
+
+func (m *prefetchManager) processTarget(ctx context.Context, target string) {
+	var task *prefetchTask
+	m.log.Infof("Prefetching OCI target: %s", target)
+
+	ctx, cancel := context.WithCancel(ctx)
+
+	m.mu.Lock()
+	task = m.tasks[target]
+	task.cancelFn = cancel
+	m.mu.Unlock()
+
+	defer cancel()
+
+	for {
+		if ctx.Err() != nil {
+			m.setResult(target, fmt.Errorf("pulling oci target %s: %w", target, ctx.Err()))
+			m.log.Warnf("Context error pulling oci target: %s", target)
+			return
+		}
+		if err := m.pull(ctx, target, task); err != nil {
+			if errors.IsRetryable(err) {
+				time.Sleep(queueNextItemAfter)
+				continue
+			}
+			m.setResult(target, fmt.Errorf("pulling oci target %s: %w", target, err))
+			return
+		}
+		// success
+		m.setResult(target, nil)
+		return
+	}
+}
+
+func (m *prefetchManager) pull(ctx context.Context, target string, task *prefetchTask) error {
+	ociType := task.ociType
+	switch ociType {
+	case OCITypeImage:
+		_, err := m.podmanClient.Pull(ctx, target, task.clientOps...)
+		if err != nil {
+			return err
+		}
+		return nil
+	case OCITypeArtifact:
+		_, err := m.podmanClient.PullArtifact(ctx, target, task.clientOps...)
+		if err != nil {
+			return err
+		}
+		return nil
+	default:
+		return fmt.Errorf("invalid oci type %s", ociType)
+	}
+}
+
+func (m *prefetchManager) setResult(image string, err error) {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	task := m.tasks[image]
+	task.err = err
+	task.done = true
+}
+
+func (m *prefetchManager) Schedule(ctx context.Context, targets []OCIPullTarget) error {
+	for _, target := range targets {
+		opts := []client.ClientOption{
+			client.Timeout(m.pullTimeout),
+		}
+		var cleanupFns []func()
+		if target.PullSecret != nil {
+			opts = append(opts, client.WithPullSecret(target.PullSecret.Path))
+			cleanupFns = append(cleanupFns, target.PullSecret.Cleanup)
+		}
+		if err := m.schedule(ctx, target.Reference, target.Type, cleanupFns, opts...); err != nil {
+			return fmt.Errorf("prefetch schedule: %w", err)
+		}
+	}
+
+	return nil
+}
+
+func (m *prefetchManager) schedule(ctx context.Context, image string, ociType OCIType, cleanupFns []func(), opts ...client.ClientOption) error {
+	m.mu.Lock()
+	if _, exists := m.tasks[image]; exists {
```

Comment:
In the event that a spec has `maxQueueSize + 1` images/artifacts to pull this check could cause an event to not get send to the download worker
 
(assume slow download times)
1. have 11 images in a spec
2. the first 10 get scheduled just fine
3. number 11 gets a task registered (line 293)
4. timer elapses and the image isn't enqueued (line 306)
5. retry eventually
6. try to schedule 11, task already exists (line 269), never sent to the worker thread.

Reply:
good catch looking

Reply:
cleaned up task